pkg/protocol/protocol_blocks.go

```diff
@@ -0,0 +1,162 @@
+package protocol
+
+import (
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/iotaledger/hive.go/ds/types"
+	"github.com/iotaledger/hive.go/ierrors"
+	"github.com/iotaledger/hive.go/lo"
+	"github.com/iotaledger/hive.go/log"
+	"github.com/iotaledger/hive.go/runtime/workerpool"
+	"github.com/iotaledger/iota-core/pkg/core/buffer"
+	"github.com/iotaledger/iota-core/pkg/model"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+// BlocksProtocol is a subcomponent of the protocol that is responsible for handling block requests and responses.
+type BlocksProtocol struct {
+	// protocol contains a reference to the Protocol instance that this component belongs to.
+	protocol *Protocol
+
+	// workerPool contains the worker pool that is used to process block requests and responses asynchronously.
+	workerPool *workerpool.WorkerPool
+
+	// droppedBlocksBuffer contains a buffer for dropped blocks that reference unsolid commitments that can be replayed
+	// at a later point in time (to make tests more reliable as we have no continuous activity).
+	droppedBlocksBuffer *buffer.UnsolidCommitmentBuffer[*types.Tuple[*model.Block, peer.ID]]
+
+	// Logger embeds a logger that can be used to log messages emitted by this chain.
+	log.Logger
+}
+
+// newBlocksProtocol creates a new blocks protocol instance for the given protocol.
+func newBlocksProtocol(protocol *Protocol) *BlocksProtocol {
+	b := &BlocksProtocol{
+		Logger:              lo.Return1(protocol.Logger.NewChildLogger("Blocks")),
+		protocol:            protocol,
+		workerPool:          protocol.Workers.CreatePool("Blocks"),
+		droppedBlocksBuffer: buffer.NewUnsolidCommitmentBuffer[*types.Tuple[*model.Block, peer.ID]](20, 100),
+	}
+
+	protocol.Constructed.OnTrigger(func() {
+		protocol.Commitments.WithElements(func(commitment *Commitment) (shutdown func()) {
+			return commitment.ReplayDroppedBlocks.OnUpdate(func(_ bool, replayBlocks bool) {
+				if replayBlocks {
+					for _, droppedBlock := range b.droppedBlocksBuffer.GetValues(commitment.ID()) {
+						b.LogTrace("replaying dropped block", "commitmentID", commitment.ID(), "blockID", droppedBlock.A.ID())
+
+						b.ProcessResponse(droppedBlock.A, droppedBlock.B)
+					}
+				}
+			})
+		})
+
+		protocol.Chains.WithElements(func(chain *Chain) func() {
+			return chain.Engine.WithNonEmptyValue(func(engineInstance *engine.Engine) (shutdown func()) {
+				return engineInstance.Events.BlockRequester.Tick.Hook(b.SendRequest).Unhook
+			})
+		})
+
+		protocol.Chains.Main.Get().Engine.OnUpdateWithContext(func(_ *engine.Engine, engine *engine.Engine, unsubscribeOnEngineChange func(subscriptionFactory func() (unsubscribe func()))) {
+			if engine != nil {
```

Comment:
As I said in a previous comment; currently during shutdown but dependening on our shutdown strategy it might turn into a superflouous check that we should get rid of over time while hardening the code,